src/analysis/special_functions/trigonometric/euler_sine_prod.lean

```diff
@@ -252,4 +252,295 @@ end
+    exact (integral_cos_pow_pos n).ne' },
+  refine this.trans _,
+  have m_nn : 0 ≤ M,
+  { replace hm := (norm_nonneg _).trans (hm (π/2) (right_mem_Icc.mpr pi_div_two_pos.le)),
+    rwa mul_nonneg_iff_left_nonneg_of_pos pi_div_two_pos at hm, },
+  conv_lhs { rw div_eq_mul_inv },
+  refine mul_le_mul_of_nonneg_left _ (div_nonneg m_nn _),
+  swap, { rw [sub_nonneg, ←nat.cast_two, nat.cast_le], exact hn.le },
+  rw inv_le,
+  { convert le_integral_cos_pow n,
+    { rw ←sqrt_inv,
+      congr' 1,
+      rw [inv_div, div_div] } },
+  { apply integral_cos_pow_pos },
+  { apply sqrt_pos_of_pos,
+    refine div_pos (mul_pos (zero_lt_two' ℝ) _) pi_pos,
+    rw [←nat.cast_add_one, nat.cast_pos],
+    linarith }
+end
+
+lemma tendsto_integral_mul_cos_pow_div
+  {f : ℝ → ℂ} (hfi : interval_integrable f volume 0 (π/2))
+  {M : ℝ} (hm : ∀ (x : ℝ), x ∈ Icc (0:ℝ) (π/2) → ‖f x - f 0‖ ≤ M * x) :
+  tendsto (λ n:ℕ, (∫ (x:ℝ) in 0..π/2, f x * cos x ^ n) / (∫ (x:ℝ) in 0..π/2, cos x ^ n))
+  at_top (𝓝 $ f 0) :=
+begin
+  have m_nn : 0 ≤ M,
+  { replace hm := (norm_nonneg _).trans (hm (π/2) (right_mem_Icc.mpr pi_div_two_pos.le)),
+    rwa mul_nonneg_iff_left_nonneg_of_pos pi_div_two_pos at hm, },
+  rw tendsto_iff_norm_tendsto_zero,
+  refine tendsto_of_tendsto_of_tendsto_of_le_of_le' tendsto_const_nhds _
+    (eventually_of_forall (λ n, norm_nonneg _))
+    ((eventually_gt_at_top 2).mp
+      (eventually_of_forall (λ n hn, abs_integral_mul_cos_pow_div_sub_le hfi hm hn))),
+  { have : tendsto (λ n:ℕ, (1:ℝ) / (n - 2) + (3:ℝ) / (n - 2) ^ 2) at_top (𝓝 0),
+    { rw (by ring : (0:ℝ) = 0 + 0),
+      refine tendsto.add (tendsto_const_nhds.div_at_top _) _,
+      { exact tendsto_at_top_add_const_right _ _ tendsto_coe_nat_at_top_at_top },
+      refine tendsto_const_nhds.div_at_top _,
+      refine (tendsto_pow_at_top two_ne_zero).comp _,
+      exact tendsto_at_top_add_const_right _ _ tendsto_coe_nat_at_top_at_top },
+    replace this := (tendsto.comp (continuous_at_id.sqrt) this).const_mul (M * sqrt (2 / π)),
+    convert tendsto.congr' _ this using 1,
+    { rw [id.def, sqrt_zero, mul_zero] },
+    rw [eventually_eq, eventually_at_top],
+    refine ⟨3, λ n hn, _⟩,
+    rw ge_iff_le at hn,
+    have : 0 < (n:ℝ) - 2,
+    { replace hn := (nat.cast_le.mpr hn : ((3:ℕ):ℝ) ≤ (n:ℝ)),
+      rw [nat.cast_bit1, nat.cast_one] at hn,
+      linarith },
+    dsimp only [id.def],
+    rw ←sq_eq_sq,
+    rw [mul_pow, mul_pow, mul_pow, sq_sqrt, sq_sqrt, sq_sqrt, div_pow],
+    { field_simp [this.ne', pi_pos.ne'],
+      ring },
+    { refine div_nonneg (mul_nonneg zero_le_two _) pi_pos.le,
+      linarith },
+    { refine add_nonneg (div_nonneg _ this.le) (div_nonneg _ (sq_pos_of_pos this).le); linarith },
+    { exact (div_pos (zero_lt_two' ℝ) pi_pos).le },
+    { refine mul_nonneg (mul_nonneg m_nn (sqrt_nonneg _)) (sqrt_nonneg _), },
+    { exact mul_nonneg (div_nonneg m_nn this.le) (sqrt_nonneg _) } },
+end
+
+lemma tendsto_integral_cos_mul_cos_pow_div (z : ℂ) :
```

Comment:
To me this is an instance of a much more general phenomenon, that has nothing to do with the cosine function: consider a space `X`, a nonnegative continuous function `c` that reaches its maximum at a single point `x_0` and is `< c (x_0) - epsilon` outside of a compact neighborhood of `x_0`, and two functions `f, g : X -> E` which are integrable and continuous at `x_0`. Then `\int f(y) c(y)^n dmu(y) / \int g(y) c(y)^n dmu(y)` tends to `f (x_0)` when `mu` is a measure supported on `s` giving positive mass to all neighborhoods of `x_0`. (All this possibly localized to a set `s` inside the space as in your application).

Reply:
Sure, that generalisation should be true. But I don't know if it is useful for anything, so I'm a little reluctant to pursue it here.

Certainly if one widens the net still further and considers any sequence of increasingly "peaky" functions $c_n(x)$, rather than just powers $c(x)^n$ of a single function, then that would include the Dirichlet and Fejer kernels, and we'd be in the territory of convergence results for Fourier series. But that would take us very far afield indeed from the original aims of this PR.

Reply:
Incidentally, Salwinski's limit formula is equivalent to some kind of averaged Fourier convergence result with binomial-coefficient weights, something like 
$$\lim_{N \to \infty} \frac{1}{2^{N}} \sum_{i = 0}^N \binom{N}{i} c_{2i-N}(f) e^{2\pi i n x}= f(x).$$
So it should already be true under a much weaker assumption; probably any $L^1$ function continuous at $x_0$ is OK.

Reply:
I don't mean to be awkward here, but: you have marked this as "awaiting author" – what are you awaiting? There's an huge panorama of possible generalisations of the theorem at hand; but generalising it far enough to be of any use outside the context of this particular argument would involve an immense amount of work.

Reply:
I really don't think it would be an immense amount of work. Probably something like 150 lines of code -- we already have a version of this theorem in `[convolution_tendsto_right](https://leanprover-community.github.io/mathlib_docs/analysis/convolution.html#convolution_tendsto_right)`, and the general proof should be barely more complicated. I would even expect the general proof to be shorter than the specific `cos x ^ n` case, because there is no temptation to do explicit computations. 

If you don't want to look at it, I can have a go. But I would be sad to merge into mathlib an argument based on computations when it hides the general phenomenon that is going on and that is useful in many situations like for the Fejer kernel.

Reply:
Which of the following are you suggesting?
- Replace `cos x ^ n` with `(c x) ^ n`, for `c` a nonnegative continuous function having a unique maximum at `x_0`. This I believe is not too difficult, but I also believe it's unlikely to be especially useful; in particular it won't cover the case of the Fejer kernel.
- Replace `cos x ^ n` with `c n x`, for `c n` _any_ sequence of functions which become increasingly concentrated at `x_0`, in some sense that's formulated broadly enough to include the Fejer kernel (and maybe the Dirichlet one as well). This seems to me to be a rather more substantial amount of work; I don't immediately see even how to precisely formulate the theorem in this generality, let alone prove it.

Reply:
I mean the second one. I had a go at it in #18327, see especially tendsto_set_integral_pow_smul_of_unique_maximum_of_is_compact_of_continuous_on there for the consequence you need in your particular case. Is the formulation good enough for you, or would you need something else?

Reply:
This certainly works in the case of the `euler_sine_prod` proof, and shortens the proof a very great deal -- great stuff!